cluster-autoscaler/proposals/plugable-provider-grpc.md

```diff
@@ -0,0 +1,604 @@
+
+In order to talk to the custom cloud provider server, this new cloud provider has to be registered
+when bootstrapping the CA.
+Consequently, the CA needs to expose new flags to specify the cloud provider and all the required properties
+to reach the remote gRPC server.
+
+A new flag, named
+`--cloud-provider-url=https://local.svc.io/mycloudprovider/server`, determines the URL to reach the custom provider implementation.
+In addition this approach reuses the existing flag that defines the name of the cloud provider using a pre-defined value `--cloud-provider=externalgrpc`
+https://github.com/kubernetes/autoscaler/tree/master/cluster-autoscaler/cloudprovider. `externalgrpc` defines the usage
+of an external cloud provider whose interface is handled by a remote gRPC service.
+
+To connect the CA core with this new external cloud provider, this approach needs to implement a new generic cloud provider
+as part of the CA core code.
+This new provider, named `ExternalGrpcCloudProvider`, makes gRPC calls to the remote functions exposed by the external cloud provider server. In other words, it forwards the calls and handle the errors analogously how done in other existing providers.
+
+Obviously, this new approach needs to use TLS to ensure a secure communication between CA and this CA provider server.
+Additional flags should be added to specify the path where to find the certificate to establish the communication to
+the external cloud provider server.
+This certificate would be mounted when deploying the cluster-autoscaler on the cluster.
+
+## User Stories
+
+### Story1
+
+When using CA only a reduced list of cloud providers are supported, if users want to use their own private cloud provider (e.g. Openstack, OpenNebula,...), they need to implement its cloud provider interface for that environment.
+This design limits the extensibility of CA and goes against certain native Kubernetes primitives.
+
+This approach aims to make CA pluggable allowing to create experimental providers without needing to change the base CA repository. This new experimental provider only needs to implement the gRPC defined operations.
+
+
+## Appendix
+
+In the following we detail all the message used in the defined actions of the `CloudProvider` GRPC Service server.
+
+### NodeGroups
+
+NodeGroups stores all node groups configured for this cloud provider.
+
+```protobuf
+
+message NodeGroupsRequest {
+  // Intentionally empty.
+}
+
+message NodeGroupsResponse {
+  // All the node groups that the cloud provider service supports. This
+  // field is OPTIONAL.
+  repeated NodeGroup nodeGroups = 1;
+}
+```
+
+### NodeGroupForNode
+
+NodeGroupForNode returns the node group for the given node, nil if the node should not be processed by cluster autoscaler, or non-nil error if such occurred. Must be implemented.
+
+```protobuf
+message NodeGroupForNodeRequest {
+  // Node group for the given node
+  k8s.io.api.core.v1.Node node = 1;
```

Comment:
Sending the whole node is probably unnecessary and likely to create performance problems (especially without caching). I would consider sending just name, provider id and maybe a few other fields that may be useful for implementation (labels and annotations?).

Reply:
It makes sense, I'll update the message properties to reflect those.

Reply:
I added the aforementioned properties, but this list could increase during the implementation phase.